src/main.rs

```diff
@@ -112,32 +115,57 @@ fn main() -> Result<()> {
                     } else {
                         vec![]
                     };
-                    for n in 0..unroll_depth {
-                        unroll_model(&mut model, n);
-                        if has_concrete_inputs {
-                            optimize_model_with_input(&mut model, &mut input_values)
-                        }
-                    }
-                    if prune {
-                        prune_model(&mut model);
-                    }
+
                     let timeout = solver_timeout.map(|&ms| Duration::from_millis(ms));
-                    match smt_solver {
-                        #[rustfmt::skip]
-                        SmtType::Generic => {
-                            optimize_model_with_solver::<none_impl::NoneSolver>(&mut model, timeout, minimize)
-                        },
-                        #[rustfmt::skip]
-                        #[cfg(feature = "boolector")]
-                        SmtType::Boolector => {
-                            optimize_model_with_solver::<boolector_impl::BoolectorSolver>(&mut model, timeout, minimize)
-                        },
-                        #[rustfmt::skip]
-                        #[cfg(feature = "z3")]
-                        SmtType::Z3 => {
-                            optimize_model_with_solver::<z3solver_impl::Z3SolverWrapper>(&mut model, timeout, minimize)
-                        },
+
+                    let mut n: usize = if stride { 1 } else { unroll_depth };
+                    loop {
+                        // TODO: Maybe we can get rid of this clone for each
+                        // iteration, which is basically required if pruning is
+                        // enabled. However, it is much faster this way than
+                        // without pruning and not cloning the model.
+                        let mut model_copy = model.clone();
+
+                        for m in 0..n {
+                            unroll_model(&mut model_copy, m);
```

Comment:
The second parameter to `unroll_model` (i.e. the `m`) is only used to construct the names of bad-state nodes and input nodes in the current unrolling step. Would it make the algorithm simpler if we were to pass the correct absolute depth at this point already? I think this could make `print_reasoning_horizon` simpler.

Reply:
Good pointer, thanks.
I adapted the implementation accordingly and `print_reasoning_horizon` did simplify significantly (as expected).